File: examples/leetcode-out/dart/10/regular-expression-matching.dart
bool isMatch(String s, String p) {
	dynamic m = s.length;
	dynamic n = p.length;
	dynamic dp = [];
	dynamic i = 0;
	while ((i <= m)) {
		dynamic row = [];
		dynamic j = 0;
		while ((j <= n)) {
			row = (row + [false]);
			j = ((j + 1)).toInt();
		}
		dp = (dp + [row]);
		i = ((i + 1)).toInt();
	}
	dp[m][n] = true;
	dynamic i2 = m;
	while ((i2 >= 0)) {
		dynamic j2 = (n - 1);
		while ((j2 >= 0)) {
			dynamic first = false;
			if ((i2 < m)) {
				if ((((_indexString(p, j2) == _indexString(s, i2))) || ((_indexString(p, j2) == ".")))) {
					first = true;
				}
			}
			dynamic star = false;
			if (((j2 + 1) < n)) {
				if ((_indexString(p, ((j2 + 1)).toInt()) == "*")) {
					star = true;
				}
			}
			if (star) {
				dynamic ok = false;
				if (dp[i2][(j2 + 2)]) {
					ok = true;
				} else {
					if (first) {
						if (dp[(i2 + 1)][j2]) {
							ok = true;
						}
					}
				}
				dp[i2][j2] = ok;
			} else {
				dynamic ok = false;
				if (first) {
					if (dp[(i2 + 1)][(j2 + 1)]) {
						ok = true;
					}
				}
				dp[i2][j2] = ok;
			}
			j2 = ((j2 - 1)).toInt();
		}
		i2 = ((i2 - 1)).toInt();
	}
	return dp[0][0];
}

void main() {
}

String _indexString(String s, int i) {
	var runes = s.runes.toList();
	if (i < 0) {
		i += runes.length;
	}
	if (i < 0 || i >= runes.length) {
		throw RangeError('index out of range');
	}
	return String.fromCharCode(runes[i]);
}

File: examples/leetcode-out/hs/10/regular-expression-matching.hs
module Main where

import Data.Maybe (fromMaybe)


forLoop :: Int -> Int -> (Int -> Maybe a) -> Maybe a
forLoop start end f = go start
  where
    go i | i < end =
            case f i of
              Just v -> Just v
              Nothing -> go (i + 1)
         | otherwise = Nothing

avg :: Real a => [a] -> Double
avg xs | null xs = 0
      | otherwise = sum (map realToFrac xs) / fromIntegral (length xs)

_indexString :: String -> Int -> String
_indexString s i =
  let idx = if i < 0 then i + length s else i
  in if idx < 0 || idx >= length s
       then error "index out of range"
       else [s !! idx]

_input :: IO String
_input = getLine


isMatch :: String -> String -> Bool
isMatch s p = fromMaybe (False) $
    (let m = length s in (let n = length p in (let dp = [] in (let i = 0 in (let dp = True in (let i2 = m in Just (((dp !! 0) !! 0))))))))
  where
    m = length s
    n = length p

main :: IO ()
main = do
    return ()
